dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -399,15 +399,121 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
   def foreachPartition[F[_]](func: Iterator[T] => Unit)(implicit F: SparkDelay[F]): F[Unit] =
     F.delay(dataset.foreachPartition(func))
 
+  /**
+    * Create a multi-dimensional cube for the current [[TypedDataset]] using the specified column,
+    * so we can run aggregation on it.
+    * See [[frameless.functions.AggregateFunctions]] for all the available aggregate functions.
+    *
+    * Differs from `Dataset#cube` by wrapping values into `Option` instead of returning `null`.
+    *
+    * apache/spark
+    */
+  def cube[K1](
+    c1: TypedColumn[T, K1]
+  ): Cube1Ops[K1, T] = new Cube1Ops[K1, T](this, c1)
+
+  /**
+    * Create a multi-dimensional cube for the current [[TypedDataset]] using the specified columns,
+    * so we can run aggregation on them.
+    * See [[frameless.functions.AggregateFunctions]] for all the available aggregate functions.
+    *
+    * Differs from `Dataset#cube` by wrapping values into `Option` instead of returning `null`.
+    *
+    * apache/spark
+    */
+  def cube[K1, K2](
+    c1: TypedColumn[T, K1],
+    c2: TypedColumn[T, K2]
+  ): Cube2Ops[K1, K2, T] = new Cube2Ops[K1, K2, T](this, c1, c2)
```

Comment:
Any reason to not use `CubeManyOps` for the arity 1 and 2 methods? Same question for the rollup methods.

Reply:
It helps compiler and with type inference:
```scala
...
val A = dataset.col[A]('a)

// Rollup1Ops, compiles, infers to TypedDataset[(Option[A], Long)]
dataset.rollup(A).agg(count())
// RollupMany, doesn't compile, infers to TypedDataset[Tuple1[Option[A]]]
dataset.rollup(A).agg(count())
// RollupMany, compiles if you specify type
dataset.rollup(A).agg(count[X1[A]])
```
I've followed `groupBy` and `select` where there are few extra methods to help with simpler use cases.

I guess the difference is that with `RelationalGroups1Ops` I can specify the `agg` return type like:
```scala
def agg[U1](c1: TypedAggregate[V, U1]): TypedDataset[(Option[K1], U1)]
```
as opposed to leaving it to macros.

Maybe it is possible to add aggregation methods and allow syntax like in `Spark`:
```scala
dataset.rollup(A).count()
```
I wasn't able to crack it yet, I'm running into similar problems.